domain/blockdevice/state/types.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/domain/blockdevice"
+)
+
+// These structs represent the persistent block device entity schema in the database.
+
+type BlockDevice struct {
+	ID string `db:"uuid"`
+
+	DeviceName     string `db:"name"`
+	Label          string `db:"label"`
+	DeviceUUID     string `db:"device_uuid"`
+	HardwareId     string `db:"hardware_id"`
+	WWN            string `db:"wwn"`
+	BusAddress     string `db:"bus_address"`
+	Size           uint64 `db:"size"`
+	FilesystemType int    `db:"filesystem_type_id"`
+	InUse          bool   `db:"in_use"`
+	MountPoint     string `db:"mount_point"`
+	SerialId       string `db:"serial_id"`
+}
+
+type FilesystemType struct {
+	ID   int    `db:"id"`
+	Name string `db:"name"`
+}
+
+type DeviceLink struct {
+	ParentUUID string `db:"block_device_uuid"`
+	Name       string `db:"name"`
+}
+
+type DeviceMachine struct {
+	BlockDeviceUUID string `db:"block_device_uuid"`
+}
+
+type BlockDevices []BlockDevice
+
+func (rows BlockDevices) toBlockDevices(deviceLinks []DeviceLink, filesystemTypes []FilesystemType) ([]blockdevice.BlockDevice, error) {
+	if n := len(rows); n != len(filesystemTypes) || n != len(deviceLinks) {
+		// Should never happen.
+		return nil, errors.New("row length mismatch")
```

Comment:
Can we look at making this error better? It is lacking context and is going to read like poo if it gets surfaced. We have context so my preference would be that we use it.

Reply:
The error is never expected to happen - this is just a defensive check. The db would be seriously messed up if we ever see this. I've expended the string slightly but really, the error will be augmented anyway with the context from the state caller - this is an internal function.